File: binutils/gdb/testsuite/gdb.fortran/vla-datatypes.f90
! Copyright 2015-2024 Free Software Foundation, Inc.
!
! This program is free software; you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation; either version 3 of the License, or
! (at your option) any later version.
!
! This program is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with this program.  If not, see <http://www.gnu.org/licenses/>.

program vla_primitives
  integer, allocatable    :: intvla(:, :, :)
  real, allocatable       :: realvla(:, :, :)
  complex, allocatable    :: complexvla(:, :, :)
  logical, allocatable    :: logicalvla(:, :, :)
  character, allocatable  :: charactervla(:, :, :)
  logical                 :: l

  allocate (intvla (11,22,33))
  allocate (realvla (11,22,33))
  allocate (complexvla (11,22,33))
  allocate (logicalvla (11,22,33))
  allocate (charactervla (11,22,33))

  l = allocated(intvla)                   ! vlas-allocated
  l = allocated(realvla)
  l = allocated(complexvla)
  l = allocated(logicalvla)
  l = allocated(charactervla)

  intvla(:,:,:) = 1
  realvla(:,:,:) = 3.14
  complexvla(:,:,:) = cmplx(2.0,-3.0)
  logicalvla(:,:,:) = .TRUE.
  charactervla(:,:,:) = char(75)

  intvla(5,5,5) = 42                      ! vlas-initialized
  realvla(5,5,5) = 4.13
  complexvla(5,5,5) = cmplx(-3.0,2.0)
  logicalvla(5,5,5) = .FALSE.
  charactervla(5,5,5) = 'X'

  ! dummy statement for bp
  l = .FALSE.                             ! vlas-modified
end program vla_primitives

File: binutils/gdb/testsuite/gdb.base/vla-datatypes.c
/* This testcase is part of GDB, the GNU debugger.

   Copyright 2014-2024 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include <stddef.h>
#define SIZE 5

struct foo
{
  int a;
};

typedef struct bar
{
  int x;
  struct foo y;
} BAR;

void
vla_factory (int n)
{
  int             int_vla[n];
  unsigned int    unsigned_int_vla[n];
  double          double_vla[n];
  float           float_vla[n];
  long            long_vla[n];
  unsigned long   unsigned_long_vla[n];
  char            char_vla[n];
  short           short_vla[n];
  unsigned short  unsigned_short_vla[n];
  unsigned char   unsigned_char_vla[n];
  struct foo      foo_vla[n];
  BAR             bar_vla[n];
  int i;

  for (i = 0; i < n; i++)
    {
      int_vla[i] = i*2;
      unsigned_int_vla[i] = i*2;
      double_vla[i] = i/2.0;
      float_vla[i] = i/2.0f;
      long_vla[i] = i*2;
      unsigned_long_vla[i] = i*2;
      char_vla[i] = 'A';
      short_vla[i] = i*2;
      unsigned_short_vla[i] = i*2;
      unsigned_char_vla[i] = 'A';
      foo_vla[i].a = i*2;
      bar_vla[i].x = i*2;
      bar_vla[i].y.a = i*2;
    }

  size_t int_size        = sizeof(int_vla);     /* vlas_filled */
  size_t uint_size       = sizeof(unsigned_int_vla);
  size_t double_size     = sizeof(double_vla);
  size_t float_size      = sizeof(float_vla);
  size_t long_size       = sizeof(long_vla);
  size_t char_size       = sizeof(char_vla);
  size_t short_size      = sizeof(short_vla);
  size_t ushort_size     = sizeof(unsigned_short_vla);
  size_t uchar_size      = sizeof(unsigned_char_vla);
  size_t foo_size        = sizeof(foo_vla);
  size_t bar_size        = sizeof(bar_vla);

  return;                                 /* break_end_of_vla_factory */
}

int
main (void)
{
  vla_factory(SIZE);
  return 0;
}
